### DISCUSSION TITLE
Achievements!

### DISCUSSION THREAD

**Discussion by TiBiBa** [hedyorg/hedy#1541] [Language Ideas]
As already proposed as an issue in #1516 we would like some way to implement achievements and the unlocking of levels. To prevent the issue overview from getting too crowded (we already have 114 open ones at this moment!) this discussion post is used to create an overview and discussion of the possible achievements as well as some implementation thoughts and issues.

## Implementations ideas
- Implement an "achievements" page for each user, which they can set to public
- Show not achieved badges grey with the % of achieved, expect for "hidden" badges
- When an achievement is reached, show an pop-up similar to modal but less intrusive
- Implement the current level / general achievement within the code-page.html
- Enable users to see their achievement in the menu bar (Not sure about this one though)
- Let users select a preferred badge to show at their profile (like a profile picture)

## Technical thoughts and issues
- Some achievement can only be tracked at the front-end whereas some can be verified at parsing
- We want to reduce server-load as much as possible
- Proposal: Create a dedicated POST function in which we handle achievements
- Call this function from either the back-end or through AJAX from the front-end
- Store achievements in .yaml files, similar to how we store adventures
- Store the reached achievement as a dict (per category) on the user in the db
- Match these at login similarly to how we match adventures with class restrictions
- When a achievement is reached, POST and re-fresh user_data from db to ensure they are aligned

## Achievements
In this subsection we discuss all possible achievements ordered by category and Hedy level. For some achievements we already thought of a catchy name. This is, of course, subject to change and might differ from the actual implementation name.

### General
- Adventure is Worthwhile (Store a program for each adventure in a level)
- Programming Protagonist (Run a successful program after an error) 
- Did you say please? (Create a program with the 'ask' command)
- Trying is Key (Use all possible commands at least once)
- Ninja Turtle (Create a program with the turtle)
- Epic Education (Join a class)

#### Programs created
- Getting Started I (Run 1 Hedy program)
- Getting Started II (Run 10 Hedy programs)
- Getting Started III (Run 50 Hedy programs)
- Getting Started IV (Run 200 Hedy programs)
- Getting Started V (Run 500 Hedy programs)

#### Storing of programs
- One to Remember I (Store 1 Hedy program)
- One to Remember II (Store 5 Hedy programs)
- One to Remember III (Store 10 Hedy programs)
- One to Remember IV (Store 25 Hedy programs)
- One to Remember V (Store 50 Hedy programs)

#### Submitting of programs
- Deadline Daredevil I (Submit 1 Hedy program)
- Deadline Daredevil II (Submit 3 Hedy programs)
- Deadline Daredevil III (Submit 10 Hedy programs) 

#### Hidden achievements
(Users will be unable to see these unless they are achieved or use GitHub and read this discussion)
- Hedy-ious (Print 10 times the same string without using a loop)
- Talk-talk-talk (Create a program with a least 5 ask keywords)
- Hedy Hacking (Create an front-end timeout due to too long execution time)
- Hedy Honor (Run a program that prints the text 'Hedy' at least once)
- Programming Panic (Try to run a faulty program 3 times in a row)
- Error or empty? (Create a program that has no output)

#### Level 1
- Use all newly introduced commands (print, ask, echo, forward, turn)
- Create a program that uses 'print', 'ask' and 'echo'
- Create 10 programs in this level
- Implement and run an adventure in this level
- Store your first program

#### Level 2
- Use all newly introduced commands (is)
- Create a program that assigns two different variables
- Create a program that assigns a variable a value twice
- Create 10 programs in this level
- Implement and run an adventure in this level

#### Level 3
- Use all newly introduced commands (at random)
- Created a program with two random commands
- Draw the turtle with randomly selected lines
- Create 10 programs in this level
- Implement and run an adventure in this level

#### Level 4
- Use all newly introduced commands (print, ask)
- Make a fault with a "forgotten quote" error
- Create 10 programs in this level
- Implement and run an adventure in this level

#### Level 5
- Use all newly introduced commands (if)
- Create a program that uses 'ask' and 'if'
- Create 10 programs in this level
- Implement and run an adventure in this level

**Comment by Felienne** [comment]
THIS IS SO COOL @TiBiBa!!

> Hedy Hacking (Create an front-end timeout due to too long execution time)

This will be hard haha because the js gives you a timeout after 3 sec and you cannot run large programs anymore. But let's see if a kid can still screw up stuff :D

**Comment by TiBiBa** [comment]
We've made a lot of progress on the achievements. For all coding progress I would like to prefer to #1543. A language-independent back-end implementation is created that depends on "achievement keys". This way the .yaml files and thereby the achievements can be translated in any language as long as the keys match. Currently it is implemented in English and Dutch. A back-end and front-end parser is written to enable us to "push" new achievement from anywhere on the user pipeline. Below an overview of the already implemented achievements within the current branch:

### General

Adventure is Worthwhile (Store a program for each adventure in a level)
**Programming Protagonist (Run a successful program after an error)**
**Did you say please? (Create a program with the 'ask' command)**
Trying is Key (Use all possible commands at least once)
**Ninja Turtle (Create a program with the turtle)**
Epic Education (Join a class)

#### Programs created (Done)

**Getting Started I (Run 1 Hedy program)
Getting Started II (Run 10 Hedy programs)
Getting Started III (Run 50 Hedy programs)
Getting Started IV (Run 200 Hedy programs)
Getting Started V (Run 500 Hedy programs)**

#### Storing of programs (Done)

**One to Remember I (Store 1 Hedy program)
One to Remember II (Store 5 Hedy programs)
One to Remember III (Store 10 Hedy programs)
One to Remember IV (Store 25 Hedy programs)
One to Remember V (Store 50 Hedy programs)**

#### Submitting of programs (Done)

**Deadline Daredevil I (Submit 1 Hedy program)
Deadline Daredevil II (Submit 3 Hedy programs)
Deadline Daredevil III (Submit 10 Hedy programs)**

#### Hidden achievements

(Users will be unable to see these unless they are achieved or use GitHub and read this discussion)

Hedy-ious (Print 10 times the same string without using a loop)
**Talk-talk-talk (Create a program with a least 5 ask keywords)**
**Hedy Hacking (Create an front-end timeout due to too long execution time)**
**Hedy Honor (Run a program that prints the text 'Hedy' at least once)**
**Programming Panic (Try to run a faulty program 3 times in a row)
Error or empty? (Create a program that has no output)**

### Levels

#### Level 1

Use all newly introduced commands (print, ask, echo, forward, turn)
Create a program that uses 'print', 'ask' and 'echo'
Create 10 programs in this level
Implement and run an adventure in this level
Store your first program

#### Level 2

Use all newly introduced commands (is)
Create a program that assigns two different variables
Create a program that assigns a variable a value twice
Create 10 programs in this level
Implement and run an adventure in this level

#### Level 3

Use all newly introduced commands (at random)
Created a program with two random commands
Draw the turtle with randomly selected lines
Create 10 programs in this level
Implement and run an adventure in this level

#### Level 4

Use all newly introduced commands (print, ask)
Make a fault with a "forgotten quote" error
Create 10 programs in this level
Implement and run an adventure in this level

#### Level 5

Use all newly introduced commands (if)
Create a program that uses 'ask' and 'if'
Create 10 programs in this level
Implement and run an adventure in this level

**Reply by Felienne** [reply]
> We've made a lot of progress on the achievements. 

Fantastic!!

**Comment by TiBiBa** [comment]
**Newly added achievements**

- Do you have a copy? (Delete a saved program)
- **Watch out! (Create a program that results in a warning)**
- **Let's focus (Turn on the programmers mode)**
- **Make Some Noise (Let your program be read aloud)**
- Next Question Please (Finish a Hedy quiz)
- Quiz Master (Get the maximum score on a quiz)
- **Double-check (Get a duplicate program name warning)**

**Reply by TiBiBa** [reply]
For some reason keeping track of deleting a program is quite hard as well perform this action using a GET instead of the expected POST with AJAX. We currently have to options to still implement this achievement:

1. Push achievements to db but show no pop-up
2. Re-write deletion of programs to a POST/AJAX combination

**Reply by Felienne** [reply]
Ah I see why this is hard! We can always skip this one for now, although I agree it is a really fun one.

I do not know *why* deletion is a GEt instead of a POST, somehow that does not seem logical to me? Can we figure out who wrote that code and ask them?

**Reply by Felienne** [reply]
> **Newly added achievements**
> 
> * Do you have a copy? (Delete a saved program)
> * **Watch out! (Create a program that results in a warning)**
> * **Let's focus (Turn on the programmers mode)**
> * **Make Some Noise (Let your program be read aloud)**
> * Next Question Please (Finish a Hedy quiz)
> * Quiz Master (Get the maximum score on a quiz)
> * **Double-check (Get a duplicate program name warning)**

Such lovely ones! I thought that maybe we can add some for _sharing_ rather than storing? And maybe one for running a program that someone else stored? (but of course we can add that to the backlist and to it later)

**Comment by TiBiBa** [comment]
Currently the achievements structure is implemented. Being language-dependent built upon .yaml files. We have a bit of a unique situation: The `/pages/achievements/en.yaml` file is **only** used for structural purposes. Namely sorting the achievements correctly and determining sections and subsections. This means that while it is called `en.yaml` it **does** not need any translation. All translated section, subsection and achievements are stored in `coursedata/achievements` making the process of translating significantly easier.

Secondly, we current perform a db call on each achievement verification check. Due to the growing nature of the achievements implementation this might get a little bit out-of-hand. As this is currently the case for when:

- We run a program
- We create an anomaly on the front-end
   - A warning
   - An error
   - A time-out
   - Modal pop-up

It might be wise to discuss some implementation changes to reduce the amount of db calls. For example, retrieve the achievements at login and store in a session. Then verify new achievements with the session instead of the db (as users might nog get achievements that frequent). If it's new: push achievement to db and re-fresh session with db call (to ensure identical sets).

**Reply by Felienne** [reply]
> Currently the achievements structure is implemented. Being language-dependent built upon .yaml files. We have a bit of a unique situation: The `/pages/achievements/en.yaml` file is **only** used for structural purposes. Namely sorting the achievements correctly and determining sections and subsections. This means that while it is called `en.yaml` it **does** not need any translation. All translated section, subsection and achievements are stored in `coursedata/achievements` making the process of translating significantly easier.

This is clear to me but to avoid confusion, maybe we should name it differently since our convention is that en.yaml is translatable content. What about `achievements.yaml` or `achievement-config.yaml`?

> Secondly, we current perform a db call on each achievement verification check. Due to the growing nature of the achievements implementation this might get a little bit out-of-hand. As this is currently the case for when:
> 
> * We run a program
> * We create an anomaly on the front-end
>   
>   * A warning
>   * An error
>   * A time-out
>   * Modal pop-up
> 
> It might be wise to discuss some implementation changes to reduce the amount of db calls. 

Yes that sounds like a plan!

> For example, retrieve the achievements at login and store in a session. Then verify new achievements with the session instead of the db (as users might nog get achievements that frequent). If it's new: push achievement to db and re-fresh session with db call (to ensure identical sets).

Storing things in the sessing instead of a DB call on each program run sounds like a good plan, however I am not entirely sure what you exactly want to store in the session yet. Do we want to store the user actions like running a program, getting an error etc. in the session and then calculate the achievements later?

**Reply by TiBiBa** [reply]
Agree, will change the structure and naming of the structural .yaml file to avoid confusion. I think the easiest way (for now) to reduce db calls it to store the achieved already achieved by the user. If we check for an achievement and it's already there we can stop the search (and don't need to call the db to check if the achievement is already there).

**Comment by TiBiBa** [comment]
**Newly added achievements**

- Ready, Set, Education! (Create a class)
- End of Semester (Delete a class)
- On Second Thoughts (Rename a class)
- My Class, My Rules (Customize a class)

**Comment by TiBiBa** [comment]
We just implemented all teacher achievements, in the process adding two more:

- Full House (Have a class with more than 20 students)
- Detention! (Remove a student from a class)

**Comment by TiBiBa** [comment]
Almost all achievements are implemented except a few. These are mainly due to having to re-write a large chunk of code before we are able to correctly check, story and display the achievements. Lastly, the `achievements.py` might need some clean-up to reduce a bit of db calls and duplicate calculations.  The following achievements are not implemented (yet):

- Trying is Key (Use all possible commands at least once)
- Do you have a copy? (Delete a saved program)
- Next Question Please (Finish a Hedy quiz)
- Quiz Master (Get the maximum score on a quiz)
- Full House (Have a class with more than 20 students)
- Detention! (Remove a student from a class)
- Hedy-ious (Print 10 times the same string without using a loop)

Also all level achievements still have to be implemented. It might be better to implement these later on as they are highly related to the "unlocking" of new levels. We should first think thoroughly about how we would best address this feature.

**Reply by Felienne** [reply]
Ok this might sound weird but... what if we changed all missing achievements into *saving* a program (Save a program with 10 lines, save a program for an adventure etc) That aligns them better with the code we already have now and makes them a lot easier to implement right?

Then, once we streamline the saving of programs to always save on run (which we already want anyway) the problem solves itself!

Or is dat a bad idea for reasons I overlooked?

**Reply by TiBiBa** [reply]
> Ok this might sound weird but... what if we changed all missing achievements into *saving* a program (Save a program with 10 lines, save a program for an adventure etc) That aligns them better with the code we already have now and makes them a lot easier to implement right?
> 
> 
> 
> Then, once we streamline the saving of programs to always save on run (which we already want anyway) the problem solves itself!
> 
> 
> 
> Or is dat a bad idea for reasons I overlooked?

That might be a nice solution! However, I think it is best to implement on the "run" functionality, even after changing the save structure. I was wondering: is there a nice function implemented in the parser that we can extract all used commands from a code snippet?

**Reply by Felienne** [reply]
> That might be a nice solution! However, I think it is best to implement on the "run" functionality, even after changing the save structure. 

Ok, you are more knowledgable of this plan than I am!

> I was wondering: is there a nice function implemented in the parser that we can extract all used commands from a code snippet? 

Not yet but I can easily make one, will add a PR for that tody

**Reply by TiBiBa** [reply]
That would be awesome! Regarding the quiz related achievements: Do you know who implemented the quiz structure? I might need so help with understanding all the code and getting it ready to "push" achievements (as we might add more achievements later on).

**Reply by Felienne** [reply]
> That would be awesome! 

I made a PR: https://github.com/Felienne/hedy/pull/1570

It is not ready yet, I need to add some tests for the higher levels, but you can merge this branch in your so you can start calling the function.

> Regarding the quiz related achievements: Do you know who implemented the quiz structure? I might need so help with understanding all the code and getting it ready to "push" achievements (as we might add more achievements later on).

@LterHaar did but I am not sure if she has bandwidth to answer questions. I think @fpereiro and @rix0rrr will also be able to answer a lot of your questions!

**Comment by TiBiBa** [comment]
Last achievement implementation update (bold is implemented):

- **Watch out! (Create a program that results in an (non-empty) warning**
- Gotta Use Them All (Use all possible commands at least once)
- **Sharing is Caring (Share a saved program)**
- **Do you have a copy? (Delete a saved program)**
- **Next Question Please (Finish a Hedy quiz)**
- **Quiz Master (Get the maximum score on a quiz)**
- **Full House (Have a class with more than 20 students)**
- **Detention! (Remove a student from a class)**
- **Hedy-ious (Print 10 times the same string without using a loop)**
